src/components/geocoder-panel.js

```diff
@@ -0,0 +1,243 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+import {SidePanelSection, PanelLabel} from './common/styled-components';
+import {FormattedMessage} from 'react-intl';
+import Geocoder from 'react-mapbox-gl-geocoder';
+import Processors from 'processors';
+import {fitBounds, addDataToMap, removeDataset} from 'actions';
+
+const GEOCODER_DATASET_NAME = 'geocoder_dataset';
+const QUERY_PARAMS = {};
+const GEO_OFFSET = 0.1;
+const ICON_LAYER = {
+  id: 'geocoder_layer',
+  type: 'icon',
+  config: {
+    label: 'Geocoder Layer',
+    color: [255, 0, 0],
+    dataId: GEOCODER_DATASET_NAME,
+    columns: {
+      lat: 'lt',
+      lng: 'ln',
+      icon: 'icon',
+      label: 'text'
+    },
+    isVisible: true,
+    hidden: true,
+    visConfig: {
+      radius: 80
+    }
+  }
+};
+
+const GeocoderPanelContent = styled.div`
+  background-color: ${props => props.theme.panelBackground};
+  padding: 0.5em 1em;
+  position: absolute;
+  top: 1em;
+  left: 50%;
+  margin-left: -20em;
+  width: 40em;
+  z-index: 10000;
+  box-sizing: border-box;
+`;
+
+const GeocoderWrapper = styled.div`
+  color: ${props => props.theme.textColor};
+  font-size: ${props => props.theme.fontSize};
+
+  .react-geocoder {
+    position: relative;
+  }
+
+  .react-geocoder input {
+    ${props => props.theme.secondaryInput};
+  }
+
+  .react-geocoder-results {
+    background-color: ${props => props.theme.panelBackground};
+    position: absolute;
+    width: 43.5em;
+  }
+
+  .react-geocoder-item {
+    ${props => props.theme.dropdownListItem};
+    ${props => props.theme.textTruncate};
+  }
+
+  .remove-layer {
+    background: transparent;
+    border: none;
+    bottom: 28px;
+    color: ${props => props.theme.textColor};
+    cursor: pointer;
+    display: inline;
+    font-size: 16px;
+    padding: 2px 8px;
+    position: absolute;
+    right: 16px;
+
+    :hover,
+    :focus,
+    :active {
+      background: transparent !important;
+      border: none;
+      box-shadow: 0;
+      color: ${props => props.theme.textColor};
+      opacity: 0.6;
+      outline: none;
+    }
+  }
+`;
+
+function generateGeocoderDataset(lat, lon, text) {
+  return {
+    data: Processors.processRowObject([
+      {
+        lt: lat,
+        ln: lon,
+        icon: 'place',
+        text
+      }
+    ]),
+    id: GEOCODER_DATASET_NAME,
+    info: {
+      hidden: true,
+      id: GEOCODER_DATASET_NAME,
+      label: GEOCODER_DATASET_NAME
+    }
+  };
+}
+
+function isValid(key) {
+  return /pk\..*\..*/.test(key);
+}
+
+export class GeocoderPanel extends Component {
+  static propTypes = {
+    isGeocoderEnabled: PropTypes.bool.isRequired,
+    mapboxApiAccessToken: PropTypes.string.isRequired
+  };
+
+  constructor() {
+    super();
```

Comment:
make `onSelected` and `removeMarker` auto bind. define [state as directly to the class](https://github.com/keplergl/kepler.gl/blob/master/src/components/common/checkbox.js#L74-L76
), then you don't need `constructor` anymore